File: src/app/types/tag.ts
export class Tag {
  constructor(
    public id: number = 0,
    public name: string = "",
    public label_id: number | null = null
  ) { }

  static fromObject(object: any): Tag {
    const { id, name, label_id } = object;
    return new Tag(id, name, label_id);
  }

  static new(label_id: number): Tag {
    return new Tag(undefined, undefined, label_id);
  }
}

export class TagCreate {
  constructor(
    public name: string,
    public label_id: number | null = null,
  ) { }
}

export class TagLabel {
  constructor(
    public id: number | null = null,
    public name: string = "",
    public category_id?: number,
    public tags: Tag[] = [],
  ) { }

  static fromObject(object: any): TagLabel {
    const { id, name, category_id, tags } = object;
    return new TagLabel(
      id,
      name,
      category_id,
      tags ? tags.map(Tag.fromObject) : []
    );
  }

  static new(category_id: number): TagLabel {
    return new TagLabel(undefined, undefined, category_id);
  }
}

File: api/database/models/Tag.py
from typing import List, Optional
from sqlmodel import SQLModel, Field, Relationship
from .ItemTagLink import ItemTagLink
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .TagLabel import TagLabel
    from .MenuItem import MenuItem

class TagBase(SQLModel):
    name: str = Field(index=True)
    label_id: Optional[int] = Field(default=None, foreign_key="taglabels.id", nullable=True)

class Tag(TagBase, table=True):
    __tablename__: str = "tags"
    id: Optional[int] = Field(default=None, primary_key=True, index=True, nullable=False)

    label: Optional['TagLabel'] = Relationship(back_populates="tags")
    menu_items: List['MenuItem'] = Relationship(back_populates="tags", link_model=ItemTagLink)

class TagRead(TagBase):
    id: int

class TagCreate(TagBase):
    pass